From d5432788a75d308b69d9c39f13e85bcb1eb13300 Mon Sep 17 00:00:00 2001
From: Emil Medve <Emilian.Medve@Freescale.com>
Date: Fri, 18 Sep 2009 14:27:02 -0500
Subject: [PATCH 037/252] P4080/DPA: Drop packets received on the error FQ

Signed-off-by: Emil Medve <Emilian.Medve@Freescale.com>
[Cleanly applied the FSL SDK 2.0.3 patch:
"kernel-2.6.30-P4080-DPA-Drop-packets-received-on-the-error.patch"]
Integrated-by: Yuri Nedel <Yuri.Nedel@windriver.com>
---
 drivers/net/dpa/dpa.c |   30 ++++++++++++++++--------------
 1 files changed, 16 insertions(+), 14 deletions(-)

diff --git a/drivers/net/dpa/dpa.c b/drivers/net/dpa/dpa.c
index aad9d46..e30fcf6 100644
--- a/drivers/net/dpa/dpa.c
+++ b/drivers/net/dpa/dpa.c
@@ -683,7 +683,8 @@ static enum qman_cb_dqrr_result ingress_rx_error_dqrr(struct qman_portal		*porta
 							struct qman_fq			*fq,
 							const struct qm_dqrr_entry	*dq)
 {
-	const struct net_device	*net_dev;
+	int			 _errno;
+	struct net_device	*net_dev;
 	const struct dpa_priv_s	*priv;
 
 	net_dev = ((struct dpa_fq *)fq)->net_dev;
@@ -692,34 +693,35 @@ static enum qman_cb_dqrr_result ingress_rx_error_dqrr(struct qman_portal		*porta
 	if (netif_msg_rx_status(priv))
 		cpu_netdev_dbg(net_dev, "-> %s:%s()\n", __file__, __func__);
 
+	net_dev->stats.rx_errors++;
+	net_dev->stats.rx_packets++;
+	net_dev->stats.rx_bytes += dpa_fd_length(&dq->fd);
+
+	BUG_ON((dq->fd.status & FM_FD_STAT_ERRORS) == 0);
+
+	cpu_netdev_err(net_dev, "%s:%hu:%s(): FD status = 0x%08x\n", __file__, __LINE__, __func__,
+		       dq->fd.status & FM_FD_STAT_ERRORS);
+
 #ifdef CONFIG_FSL_FMAN_TEST
 {
     const struct bm_buffer *bmb;
     const struct dpa_bp *dpa_bp;
-    void   *virt;
-    int	_errno;
 
     bmb = (typeof(bmb))&dq->fd;
 
     dpa_bp = dpa_bpid2pool(&priv->dpa_bp_list, bmb->bpid);
     BUG_ON(IS_ERR(dpa_bp));
 
-    virt = dpa_phys2virt(dpa_bp, bmb);
-    if (is_fman_test((void *)priv->mac_dev,
-                     FMT_RX_ERR_Q,
-                     virt,
-                     dq->fd.length20 + dq->fd.offset)) {
+	is_fman_test(priv->mac_dev, FMT_RX_ERR_Q, dpa_phys2virt(dpa_bp, bmb),
+		     dq->fd.length20 + dq->fd.offset);
+}
+#endif	/* CONFIG_FSL_FMAN_TEST */
+
         _errno = dpa_fd_release(net_dev, &dq->fd);
         if (unlikely(_errno < 0)) {
             dump_stack();
             panic("Can't release buffer to the BM during RX\n");
         }
-        return qman_cb_dqrr_consume;
-    }
-}
-#endif /* CONFIG_FSL_FMAN_TEST */
-
-	BUG();
 
 	if (netif_msg_rx_status(priv))
 		cpu_netdev_dbg(net_dev, "%s:%s() ->\n", __file__, __func__);
-- 
1.6.5.2

